src/util/ChunkedForLoop.h

```diff
@@ -0,0 +1,66 @@
+//   Copyright 2024, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#ifndef QLEVER_CHUNKEDFOREACH_H
+#define QLEVER_CHUNKEDFOREACH_H
+
+#include <algorithm>
+#include <concepts>
+#include <cstdint>
+
+namespace ad_utility {
+
+namespace detail {
+
+/// Helper function that sets the current counter to end, forcing the loop to
+/// end in the next iteration, without having to check a boolean value on each
+/// iteration. In a classic for-loop this would be the break keyword, but this
+/// doesn't work here because the code is inside a nested function.
+inline auto getSetIndexToEndAction(std::size_t& current, std::size_t end) {
+  return [&current, end]() { current = end; };
+}
+
+/// True if `Func` has a signature of `* func(size_t, const auto&)`
+/// where `*` can be any type. False otherwise.
+template <typename Func>
+constexpr bool isIteratorWithBreak =
+    std::is_invocable_v<Func, std::size_t,
+                        const decltype(getSetIndexToEndAction(
+                            std::declval<std::size_t&>(), 0))&>;
+
+/// Helper concept that allows `chunkedForLoop` to offer an action with an
+/// optional second argument in `action`.
+template <typename Func>
+concept IteratorAction =
+    std::is_invocable_v<Func, std::size_t> || isIteratorWithBreak<Func>;
+}  // namespace detail
+
+/// Helper function to run a classic for-loop from `start` to `end`, where
+/// `action` is called with the current index in the range and an optional
+/// invocable second argument that when called will cause the loop to exit in
+/// the next iteration, similar to the break keyword. `chunkOperation` is called
+/// every `CHUNK_SIZE` iteration steps, and at least a single time at the end if
+/// the range is not empty.
+template <std::size_t CHUNK_SIZE>
+inline void chunkedForLoop(std::size_t start, std::size_t end,
+                           const detail::IteratorAction auto& action,
+                           const std::invocable auto& chunkOperation) {
+  static_assert(CHUNK_SIZE != 0, "Chunk size must be non-zero");
+  using std::size_t;
+  while (start < end) {
+    size_t chunkEnd = std::min(end, start + CHUNK_SIZE);
+    while (start < chunkEnd) {
+      if constexpr (detail::isIteratorWithBreak<decltype(action)>) {
```

Comment:
No idea, to be honest I did it because the idea was to only require callers to pass the chunk size as template parameter and have everything else getting inferred.
Not sure how to achieve that when passing all the template parameters explicitly

Reply:
1. If the non-deducable parameters come first, it should still work with automatic deduction:
```
template<int x, typename T>
void f(T);
f<3>("helloBye"); // works conveniently, but f<3, int>(27.4) will incur a conversion, not intended.
```
However this allows to explicitly specify the template parameter and therefore to get funky conversions if you get it slighly wrong. Not really an issue here, but then you can use a two-phase approach to either use a struct or a lambda:
```
template <int x>
static constexpr auto f = []<typename T> (T t) {...};
f<3>("helloBye"); // exactly the syntax as above, but "other" syntax is now `f<3>.template operator()<int>(27.3)`.
```

Abseil has some fancy type trait that asserts that a template parameter was automatically deduced, but I haven't looked into how they do it.

Reply:
And in fact sonar DOES complain.